### ISSUE TITLE
Allow selecting end points of straight lines

### ISSUE THREAD

**Issue by introt** [xournalpp/xournalpp#3876]
**Is your feature request related to a problem? Please describe.**
Rearranging mind maps with lines between nodes is pretty tedious. Currently, the easiest way is to erase the lines and redraw them after moving nodes.

**Describe the solution you'd like**
Being able to select the start and/or end points of straight lines and arrows, and having lines be redrawn when one of them is moved relative to the other.

**Additional context**
![image](https://user-images.githubusercontent.com/26069787/155884986-b33cfe2f-f9fa-4862-9db7-aa4084700683.png)

Note that the line end points aren't "holding onto" any invisible node box like in many specialized graph/mind map drawing tools; they're just floating freely.

**Comment by Technius** [comment]
I think drawing diagrams is a common enough use case that it might be useful to have a dedicated "shape select" tool that is used for manipulating shape/stroke data. I've been using dedicated diagramming tools at work, and sometimes I get annoyed when it takes over 10 minutes to use the drag & drop GUI to draw a simple diagram that would've only taken 1 minute if I had access to Xournal++ pen tools...

**Comment by Technius** [comment]
Actually, I would argue that a dedicated tool or at least a special selection mode is going to be required to implement this "point selection" feature anyways.

Let's assume that we actually implement point selection is performed by the rectangular selection tool. Currently, the rectangular selection tool is designed for selecting _whole_ elements (e.g., all points in a stroke). This is fundamentally at odds with point selection, which is a sort of _partial_ element selection. If we modify the rectangular select tool to select lines that are partially in the selection region, then the selection may be much more frustrating to use when trying to pick out a specific set of strokes that are surrounded by other strokes (e.g., such as when trying to select the numerator of a handwritten fraction).

We currently don't have any dedicated tools/modes for partial element selection. Adding one will open up the opportunity to implement not only implement this feature, but also features like editing splines.

/cc @rolandlo

**Comment by introt** [comment]
I agree that a special selection mode is needed; the selection should be `"current rectangle selection" ⊆ "line shape end points"`, and as easily movable, copy-able, resizable and tilt-able as the current selections are (i.e. this tool is not for manipulating individual points).  

Other applications' stroke point selection tools, like Inkscape's "Edit paths by nodes" usually only allow selecting the points of whole paths, so we might be onto something novel here :)

**Comment by rolandlo** [comment]
> the selection should be `"current rectangle selection" ⊆ "straight line/arrow end points"` (never partial strokes!), 

Once an arrow is drawn, it is just a stroke consisting of 5 points (tail - head - side 1 of head - head - side 2 of head). Selecting all points except the point defining the tail would work fine with a "partial element selection" as suggested by @Technius

**Comment by bhennion** [comment]
> Adding one will open up the opportunity to implement not only implement this feature, but also features like editing splines.

The question for editing spline with such a tool seems tricky to me. There are a couple of cases that will require extra care:

1. The simplest one: if a spline lies outside of the selection rectangle, but one of its control point (those are usually visible) is inside, we definitely do not want to grab it with us. That's ok.

2. A more complicated situation:
![Screenshot_20220301_173007](https://user-images.githubusercontent.com/17818041/156208793-a2b7531a-98e2-47bf-8baf-c6a4e1825542.png)
So, a spline segment (with its control point). The selection box intersects the spline (we can detect that), but knowing how to distord the spline if the user drags the selection does not seem easy to me. We could cut the spline, move the selected part and fill the gap with a line segment, but that'd be quite ugly, I think.

3. If the selection box contains one of the spline's endpoint (but not both), we could imagine applying some linear transformation to the control points

4. If the selection box contains both endpoints, but not the whole spline, then we're back to a situation like 2.

**Comment by bhennion** [comment]
> > the selection should be `"current rectangle selection" ⊆ "straight line/arrow end points"` (never partial strokes!),
> 
> Once an arrow is drawn, it is just a stroke consisting of 5 points (tail - head - side 1 of head - head - side 2 of head). Selecting all points except the point defining the tail would work fine with a "partial element selection" as suggested by @Technius

It would not quite work, because the arrow head would not rotate as it should.
IMO, the arrow head should be included in the stroke style (like dashes, or newly added cap styles). The new format is an opportunity to make this change.

**Comment by bhennion** [comment]
> the selection should be `"current rectangle selection" ⊆ "straight line/arrow end points"` (never partial strokes!),

This may be a bit to restrictive: when drawing a diagram, the user does not necessarily use the ruler tool to draw a line. So the resulting "line" might still be composed of several segments (even if we can't see them).

A possible solution would be: if a stroke is long enough and straight enough, and exactly one of its endpoints is in the selection box, then we apply a linear transformation to all its points (centered at the other endpoint)

**Comment by introt** [comment]
> The question for editing spline with such a tool seems tricky to me.

I think spline editing is a separate topic, but to answer the question, if this tool were to apply to splines (which ultimately have two end points, regardless of the amount of control points, segments, etc) as well, it would work as follows:

* If the rectangle selection selects the spline as a while, this one does as well. No distortion is applied.
* Else if only one of the end points is inside the selection, that point is moved together with the selection. All other points stay where they were.
* Else the spline is not included in the selection.

I haven't seen other applications let users modify splines by selecting a subset of their points anyway.

> It would not quite work, because the arrow head would not rotate as it should.

Visualized:

![image](https://user-images.githubusercontent.com/26069787/156235961-e4dca10d-189f-47ba-a22c-33d2d8fa58cb.png)

> IMO, the arrow head should be included in the stroke style 

Agreed - and also scale in relation to the body length.

> This may be a bit to restrictive

Maybe that should be a configuration option. I personally prefer simple tools over fuzzy ones to reduce cognitive load during lectures: straight lines are a hotkey away, and can be redrawn prettier afterwards, as long as you know where they originate and end.

On that note, an eraser mode that erases whole objects only when in contact with a point of theirs (ie. ignoring straight lines except for their start and end) would be really useful. Cutting a line into two would break this selection/moving feature regardless of how many segments it was before.

To clarify: I'm not asking for a tool to edit or distort paths, strokes, etc., but one that allows me to not redraw lines when moving selections (of any shape) around. It could be a special selection tool, or a line drawing tool that creates two separately selectable/movable "end point objects" between which a connection of some sort is rendered. Could be kinda cool though if the connection wasn't a straight line, and instead automatically snaked around whatever strokes were in its way :)
